control-plane/p4RuntimeArchHandler.h

```diff
@@ -0,0 +1,384 @@
+    /// Adds architecture-specific properties for @tableBlock to the @table
+    /// Protobuf message.
+    virtual void addTableProperties(const P4RuntimeSymbolTableIface& symbols,
+                                    ::p4::config::v1::P4Info* p4info,
+                                    ::p4::config::v1::Table* table,
+                                    const IR::TableBlock* tableBlock) = 0;
+    /// Adds relevant information about @externBlock instance to the @p4info
+    /// message.
+    virtual void addExternInstance(const P4RuntimeSymbolTableIface& symbols,
+                                   ::p4::config::v1::P4Info* p4info,
+                                   const IR::ExternBlock* externBlock) = 0;
+    /// Adds relevant information about @externFunction method call - if it
+    /// needs to be exposed to the control-plane - to @p4info message.
+    virtual void addExternFunction(const P4RuntimeSymbolTableIface& symbols,
+                                   ::p4::config::v1::P4Info* p4info,
+                                   const P4::ExternFunction* externFunction) = 0;
+};
+
+/// A functor interface that needs to be implemented for each
+/// architecture-specific handler in charge of generating the appropriate P4Info
+/// message.
+struct P4RuntimeArchHandlerBuilderIface {
+    virtual ~P4RuntimeArchHandlerBuilderIface() { }
+
+    /// Called by the @ref P4RuntimeSerializer to build an instance of the
+    /// appropriate @ref P4RuntimeArchHandlerIface implementation for the
+    /// architecture, with the appropriate @refMap, @typeMap and
+    /// @evaluatedProgram.
+    virtual P4RuntimeArchHandlerIface* operator()(
+        ReferenceMap* refMap,
+        TypeMap* typeMap,
+        const IR::ToplevelBlock* evaluatedProgram) const = 0;
+};
+
+/// A collection of helper functions which can be used to implement @ref
+/// P4RuntimeArchHandlerIface for different architectures.
+namespace Helpers {
+
+/// @return an extern instance defined or referenced by the value of @table's
+/// @propertyName property, or boost::none if no extern was referenced.
+boost::optional<ExternInstance>
+getExternInstanceFromProperty(const IR::P4Table* table,
+                              const cstring& propertyName,
+                              ReferenceMap* refMap,
+                              TypeMap* typeMap,
+                              bool *isConstructedInPlace = nullptr);
+
+/// @return true if the extern instance assigned to property @propertyName for
+/// the @table was constructed in-place or outside of teh @table declaration.
+bool isExternPropertyConstructedInPlace(const IR::P4Table* table,
+                                        const cstring& propertyName);
+
+/// Visit evaluated blocks under the provided top-level block. Guarantees that
+/// each block is visited only once, even if multiple paths to reach it exist.
+template <typename Func>
+void forAllEvaluatedBlocks(const IR::ToplevelBlock* aToplevelBlock, Func function) {
```

Comment:
Isn't this just what a FlowVisitor is supposed to achieve?

Reply:
This was already heavily used in the original code so I'm going to trust that there was a reason for Seth to add this.
Do you mean `ControlFlowVisitor`? We use `forAllEvaluatedBlocks` for parsers as well, and just like I mentioned above, I'm not sure how convenient visitors are with evaluated blocks.